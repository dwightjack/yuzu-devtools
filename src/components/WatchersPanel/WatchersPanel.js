import { html } from 'lit-html';
import { component, useContext } from 'haunted';
import { noop } from '../utils';
import '../SidePanel/SidePanel';
import '../BlankSlate/BlankSlate';
import { StateContext } from '../../store/stateContext';

export default function WatchersPanel() {
  const { watchers = [], onToggleWatch = noop, onShow = noop } = useContext(
    StateContext,
  );

  return html`
    <style>
      .group {
        padding: var(--gutter) var(--gutter) calc(var(--gutter) / 2);
      }
      .group:not(:first-child) {
        border-top: 1px solid var(--color-light);
      }
      p {
        margin: calc(var(--gutter) / 2) 0;
        padding: 0;
        font-family: var(--font-monospace);
        font-size: var(--font-size-s);
        font-weight: normal;
      }
      b {
        color: var(--color-accent);
        font-weight: normal;
      }
      header {
        display: flex;
        align-items: baseline;
      }
      h3 {
        margin: 0;
        font-size: var(--font-size-m);
      }
      .btn {
        margin-left: 0.5em;
        padding: 0;
        border: 0;
        background-color: none;
        text-decoration: underline;
        color: var(--color-accent);
        font-size: var(--font-size-s);
        cursor: pointer;
      }
    </style>
    <yzdt-side-panel>
      ${watchers.length === 0
        ? html`
            <yzdt-blank-slate slot="body"
              >No watchers listening</yzdt-blank-slate
            >
          `
        : html`
            <div class="wrap" slot="body">
              ${watchers.map(({ uid, keys = [], name }) => {
                return html`
                  <div class="group">
                    <header>
                      <h3>${name}#${uid}</h3>
                      <button
                        type="button"
                        class="btn"
                        @click=${() => onShow({ uid })}
                      >
                        show
                      </button>
                    </header>
                    ${keys.map(
                      (key) => html`
                        <p>
                          <yzdt-watcher
                            key=${key}
                            uid=${uid}
                            ?watched=${true}
                            @toggle=${() => onToggleWatch(uid, key, false)}
                          ></yzdt-watcher>
                          <b>${key === '*' ? '<all>' : key}</b>
                        </p>
                      `,
                    )}
                  </div>
                `;
              })}
            </div>
          `}
    </yzdt-side-panel>
  `;
}

WatchersPanel.observedAttributes = ['name', 'uid'];
customElements.define('yzdt-watchers-panel', component(WatchersPanel));
