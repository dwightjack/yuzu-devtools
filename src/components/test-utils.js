export const mockComponent = (tag) => (props) => {
  const el = document.createElement(tag);
  Object.keys(props).forEach((key) => {
    el.setAttribute(key, props[key]);
  });
  return el;
};
