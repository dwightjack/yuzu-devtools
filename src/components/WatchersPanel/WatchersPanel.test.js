import { fireEvent } from 'dom-testing-library';
import { toHTML } from '../test-utils';
import WatchersPanel from './WatchersPanel';

describe('WatchersPanel', () => {
  const watchers = [{ uid: '_0', name: 'DEMO', keys: ['*', 'name'] }];

  test('matches default snapshot', () => {
    const { html } = toHTML(WatchersPanel, { watchers });
    expect(html).toMatchSnapshot();
  });
  test('matches empty snapshot', () => {
    const { html } = toHTML(WatchersPanel);
    expect(html).toMatchSnapshot();
  });

  test('calls onShow callback', () => {
    const onShow = jest.fn();
    const wrapper = toHTML(WatchersPanel, { watchers, onShow });
    fireEvent.click(wrapper.find('button.btn'));
    expect(onShow).toHaveBeenCalledWith({ uid: watchers[0].uid });
  });

  test('calls onToggleWatch callback on the "toggle" event of a property watcher', () => {
    const onToggleWatch = jest.fn();
    const wrapper = toHTML(WatchersPanel, { watchers, onToggleWatch });
    const watcher = wrapper.find('yzdt-watcher[key="name"][uid="_0"]');
    fireEvent(watcher, new Event('toggle'));
    expect(onToggleWatch).toHaveBeenCalledWith('_0', 'name', false);
  });
});
