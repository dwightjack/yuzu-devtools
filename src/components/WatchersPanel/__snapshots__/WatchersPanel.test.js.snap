// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WatchersPanel matches default snapshot 1`] = `
<yzdt-side-panel>
  <p slot="header"
     class="header"
  >
    <label>
      <input type="checkbox"
             name="persist-watchers"
      >
      Persist upon reloads
    </label>
  </p>
  <div class="wrap"
       slot="body"
  >
    <div class="group">
      <header>
        <h3>
          DEMO#_0
        </h3>
        <button type="button"
                class="btn"
        >
          show
        </button>
      </header>
      <p>
        <yzdt-watcher key="*"
                      uid="_0"
                      watched
        >
        </yzdt-watcher>
        <b>
          &lt;all&gt;
        </b>
      </p>
      <p>
        <yzdt-watcher key="name"
                      uid="_0"
                      watched
        >
        </yzdt-watcher>
        <b>
          name
        </b>
      </p>
    </div>
  </div>
</yzdt-side-panel>
`;

exports[`WatchersPanel matches empty snapshot 1`] = `
<yzdt-side-panel>
  <p slot="header"
     class="header"
  >
    <label>
      <input type="checkbox"
             name="persist-watchers"
      >
      Persist upon reloads
    </label>
  </p>
  <yzdt-blank-slate slot="body">
    No watchers listening
  </yzdt-blank-slate>
</yzdt-side-panel>
`;
